### PR TITLE
Add support for CMS's communication task type

### PR DESCRIPTION
**Pull request by hieplpvip**
Add support for CMS's communication task type as described here: https://cms.readthedocs.io/en/v1.4/Task%20types.html#communication

### CODE REVIEW

File: dmoj/graders/communication.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+from dmoj.executors import executors
+from dmoj.graders.standard import StandardGrader
+from dmoj.judgeenv import env, get_problem_root
+from dmoj.result import Result
+from dmoj.utils.helper_files import compile_with_auxiliary_files
+from dmoj.utils.unicode import utf8bytes, utf8text
+
+stdin_fd_flags = os.O_RDONLY
+stdout_fd_flags = os.O_WRONLY | os.O_TRUNC | os.O_CREAT
+stdout_fd_mode = stat.S_IRUSR | stat.S_IRGRP | stat.S_IROTH | stat.S_IWUSR
```
Comment (leduythuccs) [inline dmoj/graders/communication.py]: Consider to up case all of these constant

File: dmoj/graders/communication.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    """
+    Merge second_result into first_result and return first_result
+    """
+    if second_result is None:
+        raise InternalError("The second result cannot be None")
+    if first_result is None:
+        return second_result
+
+    first_result.execution_time += second_result.execution_time
+    first_result.wall_clock_time = max(first_result.wall_clock_time, second_result.wall_clock_time)
```
Comment (leduythuccs) [inline dmoj/graders/communication.py:35]: why max?

File: dmoj/graders/communication.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    first_result.max_memory += second_result.max_memory
```
Comment (leduythuccs) [inline dmoj/graders/communication.py:36]: For memory usage, it should be max instead of sum

File: dmoj/graders/communication.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        self.handler_data = self.problem.config.communication
+        if 'manager' not in self.handler_data:
+            raise InternalError('missing manager config')
+
+        self.contrib_type = self.handler_data.get('type', 'default')
+        if self.contrib_type not in contrib_modules:
+            raise InternalError('%s is not a valid contrib module' % self.contrib_type)
+
+        # We can't use self.handler_data.get('num_processes', 1) here since 0 is a falsy value
+        if 'num_processes' in self.handler_data and self.handler_data.num_processes < 1:
```
Comment (leduythuccs) [inline dmoj/graders/communication.py:55]: Can we use `self.handler_data.get('num_processes', 0) < 1`?

File: dmoj/graders/communication.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                self.problem.id, entry, aux_sources=aux_sources, defines=['-DSIGNATURE_GRADER'],
+            )
+        else:
+            raise InternalError('no valid runtime for signature grading %s found' % self.language)
+
+    def _generate_manager_binary(self):
+        files = self.handler_data.manager.files
+        if isinstance(files, str):
+            filenames = [files]
+        elif isinstance(files.unwrap(), list):
```
Comment (leduythuccs) [inline dmoj/graders/communication.py:189]: what is this :'(

File: dmoj/graders/communication.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            filenames = list(files.unwrap())
+        filenames = [os.path.join(get_problem_root(self.problem.id), f) for f in filenames]
```
Comment (leduythuccs) [inline dmoj/graders/communication.py:191]:
```py
filenames = [self.problem.problem_data[f] for f in filenames]
```
You can check the usage of problem_data in `dmoj/graders/signatures.py` (or in the `_generate_binary` above)

File: dmoj/graders/communication.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            os.chmod(self._fifo_dir[i], 0o700)
+            os.chmod(self._fifo_user_to_manager[i], 0o666)
+            os.chmod(self._fifo_manager_to_user[i], 0o666)
+
+        # Create manager processes
+        manager_args = []
+        for i in indices:
+            manager_args += [shlex.quote(self._fifo_user_to_manager[i]), shlex.quote(self._fifo_manager_to_user[i])]
+
+        self._manager_time_limit = self.num_processes * (self.problem.time_limit + 1.0)
```
Comment (leduythuccs) [inline dmoj/graders/communication.py:111]: if run time is equal to the sum run time of all processes, so why does the manager need this much time?

File: testsuite/communication_stdio_stub/manager.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  // Close FIFOs
+  fifoEncoderToManager.close();
+  fifoManagerToEncoder.close();
+  fifoDecoderToManager.close();
+  fifoManagerToDecoder.close();
+
+  if (plaintext == decrypted) {
+    cout << "1.0\n";
+    cerr << "translate:correct\n";
```
Comment (leduythuccs) [inline testsuite/communication_stdio_stub/manager.cpp]: translate:success

File: testsuite/communication_stdio/manager.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+  // Close FIFOs
+  fifoEncoderToManager.close();
+  fifoManagerToEncoder.close();
+  fifoDecoderToManager.close();
+  fifoManagerToDecoder.close();
+
+  if (plaintext == decrypted) {
+    cout << "1.0\n";
+    cerr << "translate:correct\n";
```
Comment (leduythuccs) [inline testsuite/communication_stdio/manager.cpp]: translate:success

**Review by leduythuccs — CHANGES_REQUESTED** [review]
Overall looks good, just required some changes. 
Good job

File: dmoj/graders/communication.py (hunk 4 above)
Comment (hieplpvip) [reply dmoj/graders/communication.py:55]: No, the default value should be `1`

File: dmoj/graders/communication.py (hunk 5 above)
Comment (hieplpvip) [reply dmoj/graders/communication.py:189]: lol I actually copied this from `BridgedInteractiveGrader`

File: dmoj/graders/communication.py (hunk 6 above)
Comment (hieplpvip) [reply dmoj/graders/communication.py:191]: copied from `BridgedInteractiveGrader` too :)

File: dmoj/graders/communication.py (hunk 5 above)
Comment (hieplpvip) [reply dmoj/graders/communication.py:189]: for multi-file manager. We may never need it but let's just keep it

File: dmoj/graders/communication.py (hunk 6 above)
Comment (hieplpvip) [reply dmoj/graders/communication.py:191]: `compile_with_auxiliary_files` needs filename, not file content

File: dmoj/graders/communication.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+            self.binary.populate_result(_user_proc.stderr.read(), _user_result, _user_proc)
+            final_user_result = merge_results(final_user_result, _user_result)
+
+        # The actual running time is the sum of every user process, but each
+        # sandbox can only check its own; if the sum is greater than the time
+        # limit we adjust the result.
+        if final_user_result.execution_time > self.problem.time_limit:
+            final_user_result.result_flag |= Result.TLE
+
+        result = merge_results(result, final_user_result)
```
Comment (leduythuccs) [inline dmoj/graders/communication.py]: So execution time of the manager will be added to the user execution time?

**Comment by leduythuccs** [comment]
Please add this test:
- The TL is 2.5 second, num process is 2
- A code with sleep(1) (should pass)
- A code with sleep(2) (should TLE)

**Review by leduythuccs — APPROVED** [review]
LGTM

**Pull request merged**